lib/algolia/client.rb

```diff
@@ -413,16 +429,20 @@ def thread_local_hosts(read)
       end
     end
 
-    def perform_request(session, url, method, data)
+    def perform_request(session, url, method, data, request_options)
+      hs = {}
+      extra_headers = request_options[:headers] || request_options['headers'] || {}
+      @headers.each { |key, val| hs[key.to_s] = val }
+      extra_headers.each { |key, val| hs[key.to_s] = val }
```

Comment:
You're right, I had forgotten that ruby guaranteed the iteration on hashes to be in the order of insertion (JS doesn't).